### DISCUSSION TITLE
Shouldn't the Server Action fetch have an option to include credentials?

### DISCUSSION THREAD

**Discussion by therealgilles** [vercel/next.js#72912] [App Router] [answered]
Shouldn't the Server Action fetch (below) have an option to include credentials?

From: https://github.com/vercel/next.js/blob/c50109b4468bb274c29dfc0eb5aaa906736dc4d9/packages/next/src/client/components/router-reducer/reducers/server-action-reducer.ts#L88
```TSX
  const res = await fetch('', {
    method: 'POST',
    headers: {
      Accept: RSC_CONTENT_TYPE_HEADER,
      [ACTION_HEADER]: actionId,
      [NEXT_ROUTER_STATE_TREE_HEADER]: encodeURIComponent(
        JSON.stringify(state.tree)
      ),
      ...(process.env.NEXT_DEPLOYMENT_ID
        ? {
            'x-deployment-id': process.env.NEXT_DEPLOYMENT_ID,
          }
        : {}),
      ...(nextUrl
        ? {
            [NEXT_URL]: nextUrl,
          }
        : {}),
    },
    body,
  })
```

**Comment by icyJoseph** [comment, marked as answer]
> same-origin (the default): only send and include credentials for same-origin requests.

- https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch#including_credentials

Since server actions POST to the same origin, they are included by default.

**Reply by therealgilles** [reply]
I see, thank you for the reply.

I think my problem is that I am using a real backend server during development and the cookies are associated with its subdomain. When using a server action, these cookies are not included due to the domain mismatch. I think it ought to work once deployed, but if I cannot test it out during development, it feels a bit like a gamble.

**Reply by therealgilles** [reply]
See my solution below: https://github.com/vercel/next.js/discussions/72912#discussioncomment-11425306

**Comment by therealgilles** [comment]
Follow-up question: If the server action response (from the backend server) contains cookies and a `set-cookie` header, will they be forwarded back to the client?

If not , is the only way to use an API endpoint instead?

PS: I am using an Apollo Client to make the requests to the backend. I see a `set-cookie` header on the graphql response, but it does not make it back to the browser.

**Reply by icyJoseph** [reply]
Before I answer, could you post some pseudo code to make sure we are not talking about different things.

**Reply by therealgilles** [reply]
Sure. To give you a fuller picture, the system looks like this:
>   Backend <-> NextJS Server <-> client (browser)

Let's say I have an Apollo Client for Server Actions:
```TSX
import { ApolloClient, ApolloLink, HttpLink, InMemoryCache } from '@apollo/client'
import { registerApolloClient } from '@apollo/experimental-nextjs-app-support'

export const { getClient: getClientForServerAction } = registerApolloClient(() => {
  const afterwareLink = new ApolloLink((operation, forward) => {
    return forward(operation).map(response => {
      const context = operation.getContext()
      const { response: { headers } } = context
      // The response headers are available here as `headers`, including `headers['set-cookie']`, but I don't know how to get them onto the Server Action response.
      return response
    })
  })

  const httpLink = new HttpLink({
    credentials: 'include',
    fetchOptions: { cache: 'no-cache' },
    uri: process.env.NEXT_PUBLIC_BACKEND_GRAPHQL_URL,
  })

  return new ApolloClient({
    cache: new InMemoryCache(),
    link: ApolloLink.from([afterwareLink, httpLink]),
  })
})
```

Then some kind of Server Action making a mutation request:
```TSX
'use server'

export const loginFromServer = async ({ variables }: { variables: LoginMutationVariables }) => {
  const { data } = await getClientForServerAction().mutate({
    mutation: LoginDocument,
    variables,
  })
  return { data }
}
```

and finally the call from the client-side from a login form:
```TSX
'use client'

  const handleSubmit = async (event: FormEvent<HTMLFormElement>) => {
    try {
      event.preventDefault()
      const { data } = await loginOnServer({
        variables: { userId, password, ... },
      })
      ...
  }
```

Now it's possible I'm running into a domain name or `Allow-...` header issues, as I use a real backend server in development. But I want to understand if I'm trying to do something that's not supported.

**Reply by therealgilles** [reply]
Maybe this would work in the Server Action:
```TSX
  const { data, context } = await getClientForServerAction().mutate({
    mutation: LoginDocument,
    variables,
  })
  const response = NextResponse.json({ data })
  response.headers.set('set-cookie', context.response.headers['set-cookie'])
  return response
```

Would that work?

**Reply by icyJoseph** [reply]
Yes, but not... exactly like that. 

When your server action makes itself another call to a 3rd party, that call is its own request/response pair.

Whatever headers/data, that response has, that you want to, in turn, include in the response your server action takes to the client.

Remember that a server action, is just a way for the client to invoke code on the server. 

Next.js has a way to work with cookies from within a server action: https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions-and-mutations#cookies

```ts
'use server'
 
import { cookies } from 'next/headers'
 
export async function exampleAction() {
  const cookieStore = await cookies()

 
  // Set cookie
  cookieStore.set('cookieName', cookieFromYourGQLResponse)

  return yourJSONData
}
```

**Reply by icyJoseph** [reply]
Perhaps your initial issue was also assuming that, requests made within your server action, would carry along the cookie header? You'd need to read the header, and manually pass it to your GraphQL client.

**Reply by therealgilles** [reply]
Thanks for the reply, this helps. I realized that `credentials: 'include'` has no effect on the server.

Now the question is how to get `cookieFromYourGQLResponse` back into the server action. The graphql client only returns the data. The `context` is unfortunately not sent back by the Apollo Client.

Or is there a way to access/set the NextJS cookies inside `afterwareLink`? I would try to use `await cookies()` but nothing in `afterwareLink` is async from what I can see.

**Reply by therealgilles** [reply]
Found Apollo Client `asyncMap()`, let me try it out.

**Reply by icyJoseph** [reply]
Yeah well, that's Apollo territory, and I'll gladly stay away from it.

Though you can of course, import cookies from next/header into the module that does the afterwareLink, and try to set the cookie there. It should work fine, assuming this code is only used in server actions, and that you can do asynchronous operations there.

**Reply by icyJoseph** [reply]
Also, technically speaking, I think the core question here has been answered by my first comment :) if you agree please don't forget to mark as answered.

Glad to keep on helping out, until you get it working though

**Reply by therealgilles** [reply]
`set-cookie` is a header, not a cookie. So it shouldn't be set on `cookieStore`, right? How I can update the `set-cookie` response header then?

**Reply by icyJoseph** [reply]
Nah, I meant `cookieStore set` 

The API is:

```ts

// Set cookie
  cookieStore.set('cookie-name', 'cookie-value')
```

Sorry for the confusion there, I edited my previous comment for clarity sake.

Always refer to the docs, https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions-and-mutations#cookies

**Reply by therealgilles** [reply]
If I need the cookies to be set in the browser, doesn't the client response need to contain a `set-cookie` header? Or are you saying that if I set them on the server response (as cookies), the client will get them anyway?

**Reply by icyJoseph** [reply]
Yeh, that's the promise of the cookies API.

> cookies is an async function that allows you to read the HTTP incoming request cookies in Server Component, and read/write outgoing request cookies in Server Actions or Route Handlers

- https://nextjs.org/docs/app/api-reference/functions/cookies

**Reply by therealgilles** [reply]
Ah, does `cookieStore.set()` ends up adding the `set-cookie` header to the server action response?

**Reply by icyJoseph** [reply]
You got it.

**Comment by therealgilles** [comment]
Here is the code I ended up with to use Apollo Client from React Server Actions. This only supports making one request at a time if cookies need to be set before making the next request.

```TSX
export const { getClient: getClientForServerAction } = registerApolloClient(() => {
  const addHeaders: HeadersInit = { 'Content-Type': 'application/json' }

  const authLink = setContext(async (_, { headers: prevHeaders }) => {
    // Get NextJS request cookies
    const nextRequestCookies = (await cookies()).toString()

    // Forward NextJS cookies onto GraphQL request
    const cookieHeader = {
      Cookie: `${prevHeaders?.['Cookie'] ?? ''}; ${nextRequestCookies}`,
    }

    const combinedHeaders = { ...prevHeaders, ...cookieHeader, ...addHeaders } satisfies HeadersInit

    // return the headers to the context so httpLink can read them
    return { headers: combinedHeaders }
  })

  const afterwareLink = new ApolloLink((operation, forward) => {
    return asyncMap(forward(operation), async response => {
      const {
        response: { headers: graphQLResponseHeaders },
      } = operation.getContext()

      // Extract cookies to set from graphql response `set-cookie` header
      const setCookieHeaders = graphQLResponseHeaders.getSetCookie() as ReturnType<
        Headers['getSetCookie']
      >
      const splitCookieHeaders = setCookieParser.splitCookiesString(setCookieHeaders)
      const cookiesObj = setCookieParser.parse(splitCookieHeaders)

      try {
        const cookieStore = await cookies()
        cookiesObj.forEach(cookieObj => {
          cookieStore.set(cookieObj.name, cookieObj.value, {
            path: cookieObj.path,
            expires: cookieObj.expires,
            maxAge: cookieObj.maxAge,
            domain: cookieObj.domain,
            secure: cookieObj.secure,
            httpOnly: cookieObj.httpOnly,
            sameSite: cookieObj.sameSite as boolean | 'lax' | 'strict' | 'none' | undefined,
          })
        })
      } catch (err) {
        console.log('AfterwareLink: ERROR', { operation }, err, JSON.stringify(err))
      }

      return response
    })
  })

  const httpLink = new HttpLink({
    // N.B. `credentials` has no effect on the server, it is a browser-based behavior
    // Not sure if caching applies on the server, most likely not
    fetchOptions: { cache: 'no-cache' },
    // this needs to be an absolute url, as relative urls cannot be used on the server
    uri: process.env.NEXT_PUBLIC_GRAPHQL_URL,
  })

  return new ApolloClient({
    cache: new InMemoryCache(),
    link: ApolloLink.from([logLink, authLink, afterwareLink, httpLink]),
  })
})
```
